### ISSUE TITLE
Document editors (metals) support for cross-built projects

### ISSUE THREAD

**Issue by jto** [scalacenter/scala-3-migration-guide#111]
AFAIK there's no documentation available rn about using an editor (vscode) together with metals and have support for both Scala 2.x and Scala 3 in a single project. It would be nice if that rather common use case was supported and documented.

**Comment by adpi2** [comment]
I agree that we are lacking documentation on this side. But it is a general concern about cross-built project (version and platform).

@tgodzik Don't you think it should be documented in the Metals website? This way we could add a link from here and maybe elaborate further for the context of the migration.

**Comment by tgodzik** [comment]
Not sure where to add it, we are saying [here](https://scalameta.org/metals/docs/editors/vscode.html#requirements) that Scala 3 is supported. We could add it also to overview, but the default behaviour is that Metals supports multiple versions at the same time.

We could add a mention in migration guide that multiple versions are supported and link to Metals with a comment to check exactly what is supported.

**Comment by jto** [comment]
Thanks for your replies.

Metals does support both languages out of the box but as far as I can tell, it does not support a single project having sources in both Scala 2 and Scala 3. ie. if you have macros for both languages in different folders (`src/scala-2` and `src/scala-3`).

Maybe something needs to be configured or a specific setting set in sbt, but I could not get it working.

**Comment by adpi2** [comment]
> Metals does support both languages out of the box but as far as I can tell, it does not support a single project having sources in both Scala 2 and Scala 3. ie. if you have macros for both languages in different folders (src/scala-2 and src/scala-3).

I guess you are using a `crossScalaVersions` which is a stateful way of cross building ([doc in sbt](https://www.scala-sbt.org/1.x/docs/Cross-Build.html#Cross+building+a+project+statefully)).

It means that, at any given time, sbt only knows about one scalaVersion and consequently `bloopInstall` configure Bloop for that single version. If you run `++<scalaVersion`, Bloop won't be aware of it until `bloopInstall` is run again (or `Import build` in Metals).

It is slightly better with sbt as the build server because Metals will always be in-sync with sbt when compiling. But you still need to re-import the build so that Metals is aware of the new scala-version source directory. (That can be improved in future versions with `buildChange` notification in BSP).

Using `project-matrix` works a lot better with Metals because it create one project for each scalaVersion.

So yes we should document that `project-matrix` is well supported by Metals but `crossScalaVersion` is not really.

**Comment by tgodzik** [comment]
We could try to always run `+bloopInstall`, but we would need to do a change in Bloop to generate separate build targets for each scala version. It should be simple enough though.

**Comment by jto** [comment]
Indeed I'm using `crossScalaVersions`. I would guess most cross-built projects out there do the same so it'd be great if it could "just work".

**Comment by sjrd** [comment]
While all of the above is true, and it would be nice if it worked, none of this is specific to Scala 3. If you have sources in scala-2.12 and scala-2.13, only one of those will be recognized by Metals. I don't think IntelliJ fares any better.

**Comment by jto** [comment]
I did not know that. Although while it's correct it does not make it any less of an issue. 

You are less likely to run into the issue between Scala 2.12 and 2.13 bc. they are _mostly_ compatible. OTOH If you cross-build for 2.13 and 3.0, you are almost guaranteed to run into this since the languages are so different. I worry that this lack of ergonomics will in practice hinder the work to cross-publish and in the end encourage libs maintainers to just stick to Scala 2.

**Comment by tgodzik** [comment]
@jto You are totally right. I think we should be able to work on it soon. Followed up with an issue here https://github.com/scalameta/metals/issues/2287
